### ISSUE TITLE
link verification fails

### ISSUE THREAD

**Issue by evanescente-ondine** [shinagawa-web/gomarklint#147]
Hi, URL verification fails in my setup. I just installed gomarklint with `drm@84-47-09-37-40-34 ~/D/A/NNSITE> curl -fsSL https://raw.githubusercontent.com/shinagawa-web/gomarklint/main/install.sh | sh`
My rules, close from the default, are simply:
```
{
  "default": true,
  "rules": {
    "final-blank-line": false,
    "unclosed-code-block": true,
    "empty-alt-text": true,
    "fenced-code-language": true,
    "heading-level": { "severity": "error", "minLevel": 1 },
    "duplicate-heading": true,
    "no-multiple-blank-lines": true,
    "no-setext-headings": false,
    "single-h1": false,
    "blanks-around-headings": true,
    "no-bare-urls": true,
    "no-empty-links": true,
    "no-emphasis-as-heading": true,
    "blanks-around-lists": false,
    "max-line-length": { "enabled": false, "lineLength": 80 },
    "external-link": { "enabled": true, "severity": "error", "timeoutSeconds": 5, "skipPatterns": [] }
  },
  "include": ["README.md", "testdata"],
  "ignore": [],
  "output": "text"
}
```
and it complains that apparently **all** my links are broken, but no, most are not, like   https://doi.org/10.1086/695704 or https://en.wikipedia.org/wiki/Quantitative_trait_locus or https://en.wikipedia.org/wiki/Genetic_hitchhiking.

Try with the file in attachment and the config I posted.

[ee.md](https://github.com/user-attachments/files/26944924/ee.md)

**Comment by shinagawa-web** [comment]
Thanks for the detailed report! We've reproduced and identified the root cause.

## Root Cause

Go's `net/http` client sends `User-Agent: Go-http-client/1.1` by default. Many sites — including Wikipedia and academic journal publishers (reached via doi.org redirects) — block this User-Agent and return **403 Forbidden**. Since gomarklint treats any status ≥ 400 as a broken link, all such links are incorrectly reported as unreachable.

This is especially common with sites like Wikipedia and major journal publishers (Elsevier, Wiley, APS, etc.), which sit behind CDN-level bot protection such as Cloudflare or Akamai. These services filter traffic by User-Agent at the edge, so the default Go client string tends to be turned away before the request even reaches the content server. If your Markdown files contain many links to academic or reference content, you're likely to hit this across the board.

You can verify this with the following commands:

```sh
# Wikipedia: 403 with Go's default UA
curl -s -o /dev/null -w "%{http_code}\n" \
  -A "Go-http-client/1.1" \
  https://en.wikipedia.org/wiki/Quantitative_trait_locus
# → 403

# doi.org redirects to the publisher site, which also blocks Go's UA
curl -s -o /dev/null -w "%{http_code}\n" \
  -L -A "Go-http-client/1.1" \
  https://doi.org/10.1086/695704
# → 403  (final destination: journals.uchicago.edu)

# With a descriptive UA, Wikipedia responds correctly
curl -s -o /dev/null -w "%{http_code}\n" \
  -A "gomarklint/v2 (+https://github.com/shinagawa-web/gomarklint)" \
  https://en.wikipedia.org/wiki/Quantitative_trait_locus
# → 200
```

Note: doi.org's publisher site returns 403 regardless of User-Agent, so a UA change alone is not a complete fix.

## Planned Fix

We plan to address this in two steps, tracked in a separate issue:

1. **Set a descriptive `User-Agent` header** (e.g. `gomarklint/v2 (+https://github.com/shinagawa-web/gomarklint)`) — fixes Wikipedia and similar sites that specifically block `Go-http-client/1.1`.
2. **Add a config option** (e.g. `allowedStatuses`) to treat certain HTTP status codes (such as `403`) as non-errors — for sites that block all automated clients regardless of User-Agent.

Implementation is planned for an upcoming release. In the meantime, you can work around the issue by adding patterns for affected domains to `skipPatterns` in your config.

**Comment by shinagawa-web** [comment]
@evanescente-ondine

## Update: Both planned fixes are now shipped

The two fixes described above have been implemented and merged in #183:

**1. Descriptive User-Agent**
All HTTP requests now send `User-Agent: gomarklint/v2 (+https://github.com/shinagawa-web/gomarklint)`. This resolves false positives for Wikipedia and other sites that specifically block `Go-http-client/1.1`.

**2. `allowedStatuses` config option**
A new `allowedStatuses` option lets you treat specific status codes as non-errors:

\`\`\`json
"external-link": {
  "enabled": true,
  "allowedStatuses": [403]
}
\`\`\`

This covers sites (academic publishers, social media, etc.) that return 403 regardless of User-Agent.

**Note:** 429 (Too Many Requests) is now allowed by default — rate limiting is never treated as a broken link.

---

Sites behind strict CDN bot protection may still return 403 even with the new User-Agent. For those, `allowedStatuses: [403]` or `skipPatterns` remain the recommended workaround.

**Comment by evanescente-ondine** [comment]
I installed main through
`curl -fsSL https://raw.githubusercontent.com/shinagawa-web/gomarklint/main/install.sh | sh` and the git repo alike, got version 2.12.0 in both cases,
set
```
		"external-link": {
			"enabled": true,
			"severity": "error",
			"timeoutSeconds": 5,
			"skipPatterns": ["^https://localhost"],
			"allowedStatuses": [403,302,301]
		}
```
No change whatsoever. Here's a list of links with the code status picked up by curl:
- 200: https://search.worldcat.org/title/37822067?oclcNum=37822067
	https://pmc.ncbi.nlm.nih.gov/articles/PMC5108472/
  	https://pmc.ncbi.nlm.nih.gov/articles/PMC7036703/
- 301 : https://science.org/doi/10.1126/science.abh3020
- 302 :
	https://doi.org/10.1136%2Fbmj.1.4041.1340-a
  	https://doi.org/10.1111/j.1551-2916.2006.01308.x
	https://doi.org/10.1126/science.adq7315

**Comment by shinagawa-web** [comment]
@evanescente-ondine Thanks for the follow-up. Looking at your config, it seems you're on **v2.12.0** — the `allowedStatuses` option and the User-Agent change were both added in v2.13.0 (PR #183), so they won't take effect on that version. Could you upgrade to the latest release (v3.0.5) and try again?

We tested your links against the current build. The worldcat and PMC URLs come back with 200 and pass cleanly. The `science.org` and `doi.org` links redirect to publisher sites (Science, Wiley, BMJ) that return **403** to any automated client regardless of User-Agent — that's CDN-level bot protection on their end, not a User-Agent mismatch.

With the latest version and `allowedStatuses: [403]`, all four of those links are no longer flagged.

One small note: there's no need to include `301` or `302` in `allowedStatuses`. Go's HTTP client follows redirects automatically, so the status gomarklint actually sees is the final destination's response — in this case, 403.

**Comment by nnsite** [comment]
Sorry, I didn't know how to merge a commit with git. Yes a lot more work now. I am correcting the  dangling ones, but I can see other flaws: it seems not to handle well links with ")" in them, like some DOI URLs. 
`https://doi.org/10.1016/0026-0495(67)90027-3` for instance is truncated to ` https://doi.org/10.1016/0026-0495(67` and so every such DOI links.
More exactly it creates two distinct entries,
https://doi.org/10.1016/0026-0495(67 and https://doi.org/10.1016/0026-0495
 
https://paleoanthro.org/media/journal/content/PA20170044.pdf fails too, eventhough curl shows me a code 200.

**Comment by shinagawa-web** [comment]
Both issues you reported are now fixed and released:

- **PDF links returning a non-2xx status** — Some servers (including the paleoanthropology one you mentioned) reject `HEAD` requests with 403 or 405 but serve `GET` normally. gomarklint now falls back to `GET` automatically in those cases. Fixed in #275.
- **DOI URLs with parentheses incorrectly flagged** — The URL extractor now handles balanced parentheses per the CommonMark spec, so `https://doi.org/10.1007/978-3-319-76887-8_53` and similar DOI links are parsed correctly. Fixed in #277.

---

Regarding building from source — you don't need to do that! You can install the latest release directly using whichever method suits you:

```sh
# curl (macOS / Linux)
curl -fsSL https://raw.githubusercontent.com/shinagawa-web/gomarklint/main/install.sh | sh

# Homebrew
brew install shinagawa-web/tap/gomarklint

# npm
npm install -g @shinagawa-web/gomarklint

# go install
go install github.com/shinagawa-web/gomarklint/v3@latest
```

Full installation instructions: https://shinagawa-web.github.io/gomarklint/docs/quick-start/

Thanks for the detailed report — it helped us catch both bugs!

**Comment by nnsite** [comment]
Well no, I tried the curl and go install versions and it doesn't get me the latest commits, so how am I supposed to test the commits ? With either, I still get:
> content/docs/meta.md:106: [error] Link unreachable: https://doi.org/10.1016/S0140-6736(15
 > content/docs/meta.md:106: [error] Link unreachable: https://doi.org/10.1016/S0140-6736
 > content/docs/meta.md:907: [error] Link unreachable: https://doi.org/10.1007/978-3-319-76887-8_53
> content/docs/neandertals.md:400: [error] Link unreachable: https://www.paleoanthro.org/media/journal/content/PA20170044.pdf

**Comment by shinagawa-web** [comment]
v3.0.6 is now available — the fix for parentheses in URLs (DOI, Wikipedia style) is included in this release. You can install it with:

```
go install github.com/shinagawa-web/gomarklint/v3@v3.0.6
```

DOI URLs like `https://doi.org/10.1016/S0140-6736(15)60854-4` are no longer flagged as unreachable.

**Comment by nnsite** [comment]
Perfect !! But it still fails with "https://paleoanthro.org/media/journal/content/PA20170044.pdf". Hasn't the commit been merged though ?

**Comment by shinagawa-web** [comment]
@evanescente-ondine We tested `https://paleoanthro.org/media/journal/content/PA20170044.pdf` locally with v3.0.6 and it passes cleanly.

Config used:

```json
{
  "rules": {
    "external-link": {
      "enabled": true,
      "severity": "error",
      "timeoutSeconds": 5
    }
  }
}
```

Could you try again with v3.0.6? If you're still seeing the error, could you share the exact config you're using and the full error output?

**Comment by nnsite** [comment]
The version is correct; 3.0.6. Here's my config:
```
"external-link": {
	"enabled": true,
	"severity": "error",
	"timeoutSeconds": 5,
	"skipPatterns": ["^https://localhost"],
	"allowedStatuses": [402, 403]
}
```
May I deduce that the site does not behave the same way here ? Increasing the timeout changes nothing. How can I check further ? I don't know curl enough.
